app/lib/core/Auth/AbstractLDAPAuthAdapter.php

```diff
@@ -0,0 +1,306 @@
+        if (!$this->hasRequiredGroupMembership($ps_username)) {
+            return false;
+        }
+
+        // user role and group membership syncing with directory
+        $this->syncWithDirectory($ps_username);
+
+        return true;
+    }
+    # --------------------------------------------------------------------------------
+    public function getUserInfo($ps_username, $ps_password) {
+        // ldap config
+        $vs_base_dn = $this->getConfigValue("ldap_base_dn");
+        $vs_user_ou = $this->getConfigValue("ldap_user_ou");
+        $vs_search_dn = $this->getProcessedConfigValue("ldap_user_search_dn_format", $ps_username, $vs_user_ou, $vs_base_dn);
+        $vs_search_filter = $this->getProcessedConfigValue("ldap_user_search_filter_format", $ps_username, $vs_user_ou, $vs_base_dn);
+
+        $vo_bind = $this->bindToDirectory($ps_username, $ps_password);
+        if (!$vo_bind) {
+            // wrong credentials
+            throw new LDAPException(_t("User could not be authenticated with LDAP server."));
+        }
+
+        // check group membership
+        if (!$this->hasRequiredGroupMembership($ps_username)) {
+            throw new LDAPException(_t("User is not member of at least one of the required groups."));
+        }
+
+        /* query directory service for additional info on user */
+        $vo_results = @ldap_search($this->getLinkIdentifier(), $vs_search_dn, $vs_search_filter);
+        if (!$vo_results) {
+            // search error
+            $vs_message = _t("LDAP search error: %1", ldap_error($this->getLinkIdentifier()));
+            throw new LDAPException($vs_message);
+        }
+
+        $vo_entry = ldap_first_entry($this->getLinkIdentifier(), $vo_results);
+        if (!$vo_entry) {
+            // no results returned
+            throw new LDAPException(_t("User could not be found."));
+        }
+
+        $va_attrs = ldap_get_attributes($this->getLinkIdentifier(), $vo_entry);
+
+        return array(
+            'user_name' => $ps_username,
+            'email' => $va_attrs[$this->getConfigValue("ldap_attribute_email")][0],
+            'fname' => $va_attrs[$this->getConfigValue("ldap_attribute_fname")][0],
+            'lname' => $va_attrs[$this->getConfigValue("ldap_attribute_lname")][0],
+            'active' => $this->getConfigValue("ldap_users_auto_active"),
+            'roles' => array_merge($this->getConfigValue("ldap_users_default_roles", array()), $this->getRolesToAddFromDirectory($ps_username)),
+            'groups' => array_merge($this->getConfigValue("ldap_users_default_groups", array()), $this->getGroupsToAddFromDirectory($ps_username))
+        );
+    }
+    # --------------------------------------------------------------------------------
+    public function createUserAndGetPassword($ps_username, $ps_password) {
+        // We don't create users in directories, we assume they're already there
+
+        // TODO FIXME The following is insecure!
```

Comment:
It is insecure in the following scenario:

1. User exists in LDAP and is given appropriate security group assignments to allow them access to CA.
2. User logs in to CA; the code here hashes the password and saves it in the CA database; also the roles are synced.
3. User is removed from the LDAP security group.  This has previously been all that is required to prevent that user from accessing CA.
4. Now that the `AuthenticationManager` is trying the configured adapter, and then falling back to the `CaUsers` implementation, *after the user has been removed from CA via LDAP they can still login*.  Their password still works because we saved the hash that was created here.

Actually the main source of bad security here is the fallback to `CaUsers` in `AuthenticationManager`.  Previous implementation had an explicit list of users that bypass LDAP; that was secure.

When a user is removed from the LDAP group that gives them access to CA, they should no longer be able to login to CA.